src/hotspot/share/opto/escape.cpp

```diff
@@ -4006,6 +4006,13 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
       if (n == nullptr) {
         continue;
       }
+    } else if (n->is_CallLeaf()) {
+      // Runtime calls with narrow memory input (no MergeMem node)
```

Comment:
Could we somehow assert here that we have a call with an intended narrow memory input? Directly asserting that this is an unsafe arraycopy might be too specific. But maybe we can add the following sanity check?

n->as_CallLeaf()->adr_type()->is_rawptr()

Reply:
> Could we somehow assert here that we have a call with an intended narrow memory input? Directly asserting that this is an unsafe arraycopy might be too specific. But maybe we can add the following sanity check?
> 
> ```
> n->as_CallLeaf()->adr_type()->is_rawptr()
> ```

@vnkozlov what do you think?

Reply:
If it is not Narrow memory you will get `MergeMem` node as Call's memory input which we put on`mergemem_worklist` and not processing it or its users in this part of code. We have `assert(mergemem_worklist.contains(m->as_MergeMem())` instead here.

You can add `assert(!n->in(TypeFunc::Memory)->as_MergeMem(), "only narrow memory expected here");` if you want.

Reply:
Right, we could add this assert as well for expecting a narrow memory input in general. What are your thoughts about explicitly asserting for an unsafe arraycopy when visiting this call?

Reply:
No. There is another case (DTrace runtime call) with narrow memory: [parseHelper.cpp#L54](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/parseHelper.cpp#L54)

Reply:
Since we need to integrate this P3 until tomorrow (Thursday), I'd suggest to integrate as-is and add the idea of adding an assert to the follow-up RFE.

Reply:
> No. There is another case (DTrace runtime call) with narrow memory: [parseHelper.cpp#L54](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/parseHelper.cpp#L54)

That's right. We would need to assert this case as well. But as Tobias suggested, let's move on without any additional assertions for now.